### PR TITLE
Highlight Draw Call on Mouse Hover

### PR DESCRIPTION
**Pull request by Kethku**
In progress PR which highlights the draw call under the mouse in game by subtracting 1 from the draw color.

This rewrites function calls under the current mouse position by injecting a special symbol that tells the framebufferapi to draw pixels one step brighter. With some polish I think this could be a super unique feature and would be incredibly useful to when exploring someone else's code.

The same technique could be used more broadly, but this is my first stab at it.

Also it still requires some cleanup before its production ready.

### CODE REVIEW

**Comment by Kethku** [comment]
One problem I haven't figured out how to fix is that the iframe code thinks that the game was modified when the mouse hovers over a call. I'm not sure what the cleanest way to fix that is, so I'm interested in suggestions

**Comment by gabrielflorit** [comment]
Detailed writeup here http://02credits.com/blog/day95-script-8-learnable-programming/

File: src/components/CodeEditor.js [hunk 1]
```diff
@@ -75,6 +77,74 @@ class CodeEditor extends Component {
       }
     })
 
+    // Track current mouse position in the code
+    window.addEventListener('mousemove', e => {
```
Comment (gabrielflorit) [inline src/components/CodeEditor.js]: Should we limit the event listener to the code editor dom node? Or maybe even let react handle the listening?

File: src/components/CodeEditor.js [hunk 2]
```diff
@@ -140,10 +210,10 @@ class CodeEditor extends Component {
     // If it's a number,
     if (token && token.type === 'number') {
       // clear out the previous mark,
-      this.mark && this.mark.clear()
+      this.sliderMark && this.sliderMark.clear()
 
       // and mark this token.
-      this.mark = this.codeMirror.markText(
+      this.sliderMark = this.codeMirror.sliderMarkText(
```
Comment (gabrielflorit) [inline src/components/CodeEditor.js]: I agree with renaming `mark` to `sliderMark`, but we need to leave `this.codeMirror.markText` as is. This was probably a simple typo.

File: src/reducers/game.js [hunk 3]
```diff
@@ -53,7 +53,13 @@ const game = handleActions(
     }),
 
     [actionTypes.FETCH_GIST_SUCCESS]: (state, action) =>
-      parseGistGame(action.payload)
+      parseGistGame(action.payload),
+
+    [actionTypes.SET_CALL_UNDER_MOUSE]: (state, { payload }) => ({
```
Comment (gabrielflorit) [inline src/reducers/game.js]: Is there a need for a `CLEAR_CALL_UNDER_MOUSE`? Why not?

**Comment by gabrielflorit** [comment]
> One problem I haven't figured out how to fix is that the iframe code thinks that the game was modified when the mouse hovers over a call. I'm not sure what the cleanest way to fix that is, so I'm interested in suggestions

1) Is this a problem? I mean, when you hover over a call, how often does the iframe code re-eval? Is it just once, just like when you type an extra character? Or is it 60fps for as long as you're hovering?
2) If it is a problem, one simple (and probably not clever) way is to add a boolean to the payload `Output.js` is sending to `Iframe.js`, with a message like `DONOTEVAL`.

File: src/components/CodeEditor.js (hunk 2 above)
Comment (Kethku) [reply src/components/CodeEditor.js]: ah yup my bad

**Comment by Kethku** [comment]
Its only a problem in that its weird. I don't think it could cause any functional issues as when you move your mouse off of the text editor to go save the cart, the injected code would be gone. That said, its its a bit of an oddity, so it should probably be fixed.

As for how often does it happen, I think it only happens when the mouse moves, and when the call under the mouse actually changes (I think... in my testing outside of the project I implemented the change but I need to double check thats actually happening here), so performance wise its no worse this way than before. Eval only happens when the code changes, so 60 fps is kinda unrelated maybe? I don't remember for sure.

To really fix it I think we need to set a "dontSave" flag in the redux state and have the changed indicator check the "dontSave" flag before updating its UI. Maybe that would be a more principled solution?

I'd be happy to work on these and update my PR, I just wanted feedback first.

File: src/components/CodeEditor.js (hunk 1 above)
Comment (Kethku) [reply src/components/CodeEditor.js]: yeah thats probably a good idea. I'll make the change

Comment (gabrielflorit) [reply src/components/CodeEditor.js]: Oh also we should remove it on `componentWillUnmount`.

**Comment by Kethku** [comment]
I'm confused. I think it has to be evaled because thats how the graphics get changed to reflect the hovered function. The struggle is that the current path for evaled code also updates what code would be stored in a gist. Thats the bit I don't know how to tease apart.

File: src/components/CodeEditor.js (hunk 1 above)
Comment (Kethku) [reply src/components/CodeEditor.js]: ah yes. Good point.

**Comment by gabrielflorit** [comment]
Oh of course. I see.

File: src/reducers/game.js (hunk 3 above)
Comment (Kethku) [reply src/reducers/game.js]: I use calling setCallUnderMouse with null to clear it. I will change this to a specialized clear command

**Pull request merged**